File: src/main/python/net/coderextreme/data/ExtrusionHeart.py
print('<!--')
import x3d
print('-->')
X3D0 = x3d.X3D()
X3D0.profile = "Immersive"
X3D0.version = "4.1"
head1 = x3d.head()
meta2 = x3d.meta()
meta2.name = "title"
meta2.content = "ExtrusionHeart.x3d"

head1.children.append(meta2)
meta3 = x3d.meta()
meta3.name = "description"
meta3.content = "Simple extrusion of a Valentine heart."

head1.children.append(meta3)
meta4 = x3d.meta()
meta4.name = "creator"
meta4.content = "Class participants in course Introduction to VRML/X3D."

head1.children.append(meta4)
meta5 = x3d.meta()
meta5.name = "created"
meta5.content = "14 February 2001"

head1.children.append(meta5)
meta6 = x3d.meta()
meta6.name = "modified"
meta6.content = "Mon, 09 Feb 2026 12:23:01 GMT"

head1.children.append(meta6)
meta7 = x3d.meta()
meta7.name = "identifier"
meta7.content = "https://www.web3d.org/x3d/content/examples/Basic/course/ExtrusionHeart.x3d"

head1.children.append(meta7)

X3D0.head = head1
Scene8 = x3d.Scene()
Viewpoint9 = x3d.Viewpoint()
Viewpoint9.description = "Extrusion Heart"
Viewpoint9.position = [0,-4,0]
Viewpoint9.orientation = [1,0,0,1.57]

Scene8.children.append(Viewpoint9)
Transform10 = x3d.Transform()
Transform10.translation = [0,-0.5,0]
Shape11 = x3d.Shape()
Appearance12 = x3d.Appearance()
Material13 = x3d.Material()
Material13.diffuseColor = [0.8,0.3,0.3]

Appearance12.material = Material13

Shape11.appearance = Appearance12
Extrusion14 = x3d.Extrusion()
Extrusion14.solid = False
Extrusion14.creaseAngle = 3.14159
Extrusion14.crossSection = [(0, 0.8),(0.2, 1),(0.7, 0.95),(1, 0.5),(0.8, 0),(0.5, -0.3),(0, -0.7),(-0.5, -0.3),(-0.8, 0),(-1, 0.5),(-0.7, 0.95),(-0.2, 1),(0, 0.8)]
Extrusion14.scale = [(0.01, 0.01),(0.8, 0.8),(1, 1),(0.8, 0.8),(0.01, 0.01)]
Extrusion14.spine = [(0, 0, 0),(0, 0.1, 0),(0, 0.5, 0),(0, 0.9, 0),(0, 1, 0)]

Shape11.geometry = Extrusion14

Transform10.children.append(Shape11)

Scene8.children.append(Transform10)

X3D0.Scene = Scene8
f = open("../data/ExtrusionHeart.new.python.x3d", mode="w", encoding="utf-8")
f.write(X3D0.XML())
f.close()
f = open("../data/ExtrusionHeart.new.python.x3dv", mode="w", encoding="utf-8")
f.write(X3D0.VRML())
f.close()
f = open("../data/ExtrusionHeart.new.python.x3dj", mode="w", encoding="utf-8")
f.write(X3D0.JSON())
f.close()

File: src/main/go/net/coderextreme/data/ExtrusionHeart.go

package main

import (
	"encoding/xml"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"

	"gox3d/x3d" // Assumes your module is named x3d-go-tester

	// Imports for the CGo-based libxml2 wrapper
	"github.com/lestrrat-go/libxml2"
	"github.com/lestrrat-go/libxml2/xsd"
)

// ... (Helper functions remain the same) ...
func stringPtr(s string) *string { return &s }
func floatPtr(f float32) *float32 { return &f }
func doublePtr(d float64) *float64 { return &d }
func boolPtr(b bool) *bool       { return &b }
func int32Ptr(i int32) *int32    { return &i }

func downloadSchemaIfNotExists(url, filepath string) error {
	// ... (function is correct, no changes) ...
	if _, err := os.Stat(filepath); err == nil {
		fmt.Printf("Schema file '%s' already exists. Skipping download.\n", filepath)
		return nil
	}
	fmt.Printf("Downloading X3D schema from %s to %s...\n", url, filepath)
	resp, err := http.Get(url)
	if err != nil { return fmt.Errorf("failed to download schema: %w", err) }
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK { return fmt.Errorf("bad status: %s", resp.Status) }
	out, err := os.Create(filepath)
	if err != nil { return fmt.Errorf("failed to create schema file: %w", err) }
	defer out.Close()
	_, err = io.Copy(out, resp.Body)
	if err != nil { return fmt.Errorf("failed to save schema: %w", err) }
	fmt.Println("Schema downloaded successfully.")
	return nil
}

// Updated validation function to print detailed errors
func validateXMLWithSchema(xmlData []byte, schemaPath string) error {
	schema, err := xsd.ParseFromFile(schemaPath)
	if err != nil { return fmt.Errorf("failed to parse schema file %s: %w", schemaPath, err) }
	defer schema.Free()
	doc, err := libxml2.Parse(xmlData)
	if err != nil { return fmt.Errorf("failed to parse XML data: %w", err) }
	defer doc.Free()

	validationErr := schema.Validate(doc)
	if validationErr != nil {
		errorLog := ""
		if schemaErrors, ok := validationErr.(xsd.SchemaValidationError); ok {
			for i, e := range schemaErrors.Errors() {
				errorLog += fmt.Sprintf("Error %d: %s\n", i+1, e.Error())
			}
			return fmt.Errorf("schema validation failed with %d errors:\n%s", len(schemaErrors.Errors()), errorLog)
		}
		return fmt.Errorf("schema validation failed: %w", validationErr)
	}
	return nil
}

func main() {
	fmt.Println("--- Building and Testing an X3D Scene in Go ---")

	const schemaURL = "https://www.web3d.org/specifications/x3d-4.0.xsd"
	const schemaFilename = "x3d-4.0.xsd"
	if err := downloadSchemaIfNotExists(schemaURL, schemaFilename); err != nil {
		log.Fatalf("Could not prepare schema file: %v", err)
	}
    sceneRoot := &x3d.X3D{
        Profile: stringPtr("Immersive"),
        Version: stringPtr("4.0"),
        Head: &x3d.Head{
            Metas: []*x3d.Meta{
                &x3d.Meta{
                    Name: stringPtr("title"),
                    Content: stringPtr("ExtrusionHeart.x3d"),
            },
            &x3d.Meta{
                Name: stringPtr("description"),
                Content: stringPtr("Simple extrusion of a Valentine heart."),
            },
            &x3d.Meta{
                Name: stringPtr("creator"),
                Content: stringPtr("Class participants in course Introduction to VRML/X3D."),
            },
            &x3d.Meta{
                Name: stringPtr("created"),
                Content: stringPtr("14 February 2001"),
            },
            &x3d.Meta{
                Name: stringPtr("modified"),
                Content: stringPtr("Mon, 08 Sep 2025 00:43:52 GMT"),
            },
            &x3d.Meta{
                Name: stringPtr("identifier"),
                Content: stringPtr("https://www.web3d.org/x3d/content/examples/Basic/course/ExtrusionHeart.x3d"),
            },
            },
        },
        Scene: &x3d.Scene{
            Children: []x3d.X3DChildNode{
                &x3d.Viewpoint{
                    Description: stringPtr("Extrusion Heart"),
                    Position: &x3d.SFVec3f{0.0, -4.0, 0.0},
                    Orientation: &x3d.SFRotation{1.0, 0.0, 0.0, 1.57},
                },
                &x3d.Transform{
                        Translation: &x3d.SFVec3f{0.0, -0.5, 0.0},
                    Children: []x3d.X3DNode{
                        &x3d.Shape{
                            Appearance: &x3d.Appearance{
                                Material: &x3d.Material{
                                    DiffuseColor: &x3d.SFColor{0.8, 0.3, 0.3},
                                },
                            },
                            Geometry: &x3d.Extrusion{
                                Solid: boolPtr(false),
                                CreaseAngle: floatPtr(3.14159),
                                CrossSection: &x3d.MFVec2f{[2]float32{0.0,0.8},[2]float32{0.2,1.0},[2]float32{0.7,0.95},[2]float32{1.0,0.5},[2]float32{0.8,0.0},[2]float32{0.5,-0.3},[2]float32{0.0,-0.7},[2]float32{-0.5,-0.3},[2]float32{-0.8,0.0},[2]float32{-1.0,0.5},[2]float32{-0.7,0.95},[2]float32{-0.2,1.0},[2]float32{0.0,0.8}},
                                Scale: &x3d.MFVec2f{[2]float32{0.01,0.01},[2]float32{0.8,0.8},[2]float32{1.0,1.0},[2]float32{0.8,0.8},[2]float32{0.01,0.01}},
                                Spine: &x3d.MFVec3f{[3]float32{0.0,0.0,0.0},[3]float32{0.0,0.1,0.0},[3]float32{0.0,0.5,0.0},[3]float32{0.0,0.9,0.0},[3]float32{0.0,1.0,0.0}},
                            },
                        },
                    },
                },
            },
        },
    }


	fmt.Println("\n--- Validating the generated scene (internal logic) ---")
	err := sceneRoot.Scene.Validate()
	if err != nil {
		log.Fatalf("Scene validation failed: %v", err)
	}
	fmt.Println("✅ Internal scene validation successful!")
	fmt.Println("\n--- Serializing scene to XML ---")
	output, err := xml.MarshalIndent(sceneRoot, "", "  ")

	if err != nil {
		log.Fatalf("XML Marshaling failed: %v", err)
	}
	/*
	fmt.Println("\n--- Validating XML against X3D 4.0 Schema (using libxml2) ---")
	err = validateXMLWithSchema(output, schemaFilename)
	if err != nil {
		fmt.Printf("--- Invalid Generated XML ---\n%s\n---------------------------\n", string(output))
		log.Fatalf("Schema validation failed for generated XML: %v", err)
	}
	fmt.Println("✅ XML is valid against the X3D 4.0 schema!")
	*/
	filename := "../data/ExtrusionHeart.new.go.x3d"
	file, err := os.OpenFile(filename, os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		fmt.Printf("Error opening file: %v\n", err)
		return
	}
	defer file.Close() // Ensure the file is closed when the function exits

	// Write the string content to the file
	header := "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<!DOCTYPE X3D PUBLIC \"ISO//Web3D//DTD X3D 4.0//EN\" \"https://www.web3d.org/specifications/x3d-4.0.dtd\">\n"
	_, err = file.WriteString(header)
	if err != nil {
		fmt.Printf("Error writing header to file: %v\n", err)
		return
	}

	_, err = file.Write(output)
	if err != nil {
		fmt.Printf("Error writing bytes to file: %v\n", err)
		return
	}

	fmt.Printf("\n✅ Scene successfully written to %s\n", filename)
	fmt.Printf("   You can view the generated XML below:\n\n%s\n", string(output))
}